S3/S3.py

```diff
@@ -870,43 +874,7 @@ def urlencode_string(self, string, urlencoding_mode = None):
             ## Don't do any pre-processing
             return string
 
-        encoded = ""
-        ## List of characters that must be escaped for S3
-        ## Haven't found this in any official docs
-        ## but my tests show it's more less correct.
-        ## If you start getting InvalidSignature errors
-        ## from S3 check the error headers returned
-        ## from S3 to see whether the list hasn't
-        ## changed.
-        for c in string:    # I'm not sure how to know in what encoding
-                    # 'object' is. Apparently "type(object)==str"
-                    # but the contents is a string of unicode
-                    # bytes, e.g. '\xc4\x8d\xc5\xafr\xc3\xa1k'
-                    # Don't know what it will do on non-utf8
-                    # systems.
-                    #           [hope that sounds reassuring ;-)]
-            o = ord(c)
-            if (o < 0x20 or o == 0x7f):
-                if urlencoding_mode == "fixbucket":
-                    encoded += "%%%02X" % o
-                else:
-                    error(u"Non-printable character 0x%02x in: %s" % (o, string))
-                    error(u"Please report it to s3tools-bugs@lists.sourceforge.net")
-                    encoded += replace_nonprintables(c)
-            elif (o == 0x20 or  # Space and below
-                o == 0x22 or    # "
-                o == 0x23 or    # #
-                o == 0x25 or    # % (escape character)
-                o == 0x26 or    # &
-                o == 0x2B or    # + (or it would become <space>)
-                o == 0x3C or    # <
-                o == 0x3E or    # >
-                o == 0x3F or    # ?
-                o == 0x60 or    # `
-                o >= 123):      # { and above, including >= 128 for UTF-8
-                encoded += "%%%02X" % o
-            else:
-                encoded += c
+        encoded = quote_plus(string, safe="~/")
```

Comment:
For "decode_from_s3()" and "encode_to_s3()", I agree.
But not for the rest.

I'm not really expert, but my understanding is the following:
Sadly, contrary to his name, getfilesystemencoding doesn't return the encoding of a targeted filesystem (like the one we would like to write to), but the system config "CODESET".
Also, to use it we have to take care to use default encoding because of that:
"Return the name of the encoding used to convert Unicode filenames into system file names, or None if the system default encoding is used. "

And for getdefaultencoding, it is not clear what it is useful to, for python I think that it is hardly fixed to ASCII.

From the source code of getpreferredencoding, it looks like that it also based on CODESET, so should return the same thing as getfilesystemencoding. But, maybe it could be modified by user env vars preferences that is better I guess.
And I don't think that there is a difference to do between the "console" encoding and the one to access files. Because one could assume that they should be compatible.

For the config value, please don't plan to remove it as it is valuable in my case.
To give you an indication of my current use case:
- I have a system that is setup in ASCII C encoding. But the system is able to support UTF-8 and so are the user data mounted filesystems.
- s3cmd has no way to be able to guess the right encoding to use, but it is ok for me to set "UTF-8" in the configuration and so everything will works ok.